[CUDA Host Allocator] Add support of CudaHostRegister

banitag1 opened the pull request:
Summary: This diff adds another option to create cuda pinned memory using cudaHostRegister.

Differential Revision: D45843715

zdevito commented on test/test_cuda.py:
```diff
@@ -76,6 +76,13 @@ def tearDown(self):
         super().tearDown()
 
+    def test_pinned_memory_with_cudaregister(self):
+        os.environ["PYTORCH_CUDA_ALLOC_CONF"] = "pinned_use_cuda_host_register:True,pinned_num_register_threads:8"
```
This should be `torch.cuda.memory._set_allocator_settings` otherwise it won't have an effect. Make sure to set it back after the test.

banitag1 committed:
[CUDA Host Allocator] Add support of CudaHostRegister (#108488)

Summary:

This diff adds another option to create cuda pinned memory using cudaHostRegister to avoid large lock wait time with cudaHostAlloc.

Reviewed By: zdevito

Differential Revision: D45843715
